wallet-new/src/Cardano/Wallet/API/V1/Errors.hs

```diff
@@ -52,6 +52,7 @@ data WalletError =
     | SomeOtherError { weFoo :: !Text, weBar :: !Int }
     | MigrationFailed { weDescription :: !Text }
     | JSONValidationFailed { weValidationError :: !Text }
+    | GenericError { weMsg :: !Text }
```

Comment:
I would prefer to see `UnkownError SomeException` instead -- that would let us retain more information about the error up until the serialization boundary. I strongly dislike `Text` and friends for errors, as they are completely opaque and any other error can be serialized to them later.

Reply:
I see you point. However, we are using `WalletError` at _the serialization boundary_ . That means we have to serialize / deserialize `WalletError` (incl. `UnkownError SomeException`) in that case. Not least to deal with `JSend`. How would you do that? 

We have already discussed about that issue ^ in our team today. Please check https://input-output-rnd.slack.com/archives/C819S481Y/p1520592997000412 and all following comments there.

Reply:
Serializing works much the same:

```haskell
Prelude> import Control.Exception
Prelude Control.Exception> :t toException (userError "oops")
toException (userError "oops") :: SomeException
Prelude Control.Exception> show (toException (userError "oops"))
"user error (oops)"
```

When we have a `SomeException` we can always show it, and it'll delegate to the `displayException` (typically `show`) function on the `Exception` class.

---

But there's more! Suppose we have a `GenericError` on our hands -- we can't do anything with it. Our hands are tied. Alas. We can print the message. If we wanted to parse it, we could, but that's a ton of work. If we have an `UnknownError`, then we can use the `fromException :: Exception e => SomeException -> Maybe e` method to test potential causes and handle them separately. This makes `WalletError` here useful at the API boundary *and also* at other points in the stack.

Reply:
@parsonsmatt I finally had the chance of playing with the code and I can now see the lost in translation with @sectore : when we originally spoke, I was under the impression that generating a valid generic instance for `SomeException` was not possible, but turns out that [deriveGeneric](http://hackage.haskell.org/package/generics-sop-0.3.2.0/docs/Generics-SOP-TH.html#v:deriveGeneric) from `Generics.SOP.TH` can do that just as well. I think the compilation errors Jens was experiencing are completely fixable if we:

* We remove the `Eq` constraint on `WalletError` or define an orphan for `SomeException` (which would be better to avoid, in which case we might want to rewrap `SomeException` in a `newtype`)

* We remove completely `FromJSON WalletError`, as I think that originally we added this extra instance to be able to write roundtrips in the specs, but practically we should be ever in a position to parse a `WalletError` as input. At least that's a use case I cannot anticipate on top of my head.

Addressing the two above should be enough to have our cake and eat it too: I was, in fact, able to get a clean build relatively quickly. Let me know if I'm being mental.

Reply:
We'd need to parse a `WalletError` in the servant-client. Alternatively we can just rely on the `ServantError` type and not have fine-grained error handling on clients.

Reply:
The missing Eq and FromJSON instances are pretty big deals. I'm happy to leave this as text for now.